### PR TITLE
add oper.d

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Comment by WalterBright** [comment]
@ibuclaw gdc seems to be having some trouble:
---
backend/ty.d:297:31: error: variable TYMAX extern symbols cannot have initializers
 extern __gshared ubyte[TYMAX] _tyrelax;
                               ^
backend/oper.d:309:31: error: variable OPMAX extern symbols cannot have initializers
 extern __gshared ubyte[OPMAX] optab1, optab2, optab3;
                               ^
make: *** [dmd] Error 1
---
Looks like a bug in gdc. Any ideas?

**Comment by WalterBright** [comment]
Looks like the patch I made works.

File: src/backend/oper.d [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+extern (C)
+{
+    static if (__VERSION__ > 2066)
+    {
+        extern __gshared ubyte[OPMAX] optab1, optab2, optab3;
+        extern __gshared ubyte[OPMAX] opcost;
+    }
+    else
+    {
+        extern __gshared ubyte[179] optab1, optab2, optab3;
```
Comment (MartinNowak) [inline src/backend/oper.d]: static assert(OPMAX == 179);

Comment (MartinNowak) [inline src/backend/oper.d]: See https://github.com/dlang/dmd/pull/6034/commits/962ef0614645ff313a4e69120ceb57eaad2ba5e4#diff-c787edcd24b34ddec8409246ca1c5e4bR296 for a slightly nicer workaround.

```d

 // Workaround 2.066.x bug by resolving the OPMAX value before using it as dimension.
 static if (__VERSION__ <= 2066):
     private enum computeEnumValue = OPMAX;
```

Comment (WalterBright) [inline src/backend/oper.d]: Nice! Incorporated.

File: src/posix.mak [hunk 2]
```diff
@@ -145,7 +145,7 @@ CXXFLAGS += \
     -std=gnu++98
 endif
 # Default D compiler flags for all source files
-DFLAGS=
+DFLAGS= version=MARS
```
Comment (yebblies) [inline src/posix.mak]: `-version`?

**Comment by ibuclaw** [comment]
@WalterBright - I must have remembered things wrong as I thought the compiler used by TravisCI was newer.  When a new packaged build is available I'll revert the workarounds here.

File: src/posix.mak (hunk 2 above)
Comment (WalterBright) [inline src/posix.mak]: fixed

File: src/backend/oper.d [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+ *      OTrel           == != < <= > >= operators
+ *      OTrel2          < <= > >= operators
+ *      OTdef           definition operator (assign call post asm)
+ *      OTae            potential common subexpression operator
+ *      OTboolnop       operation is a nop if boolean result is desired
+ */
+
+// Workaround 2.066.x bug by resolving the TYMAX value before using it as dimension.
+static if (__VERSION__ <= 2066):
+    private enum computeEnumValue = TYMAX;
```
Comment (MartinNowak) [inline src/backend/oper.d]: OPMAX

Comment (WalterBright) [inline src/backend/oper.d]: geez, I can't do anything right today :-(

File: src/backend/oper.d [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+module ddmd.backend.oper;
+
+extern (C++):
+@nogc:
+nothrow:
+
+alias int OPER;
+enum
+{
+        OPunde,                 // place holder for undefined operator
```
Comment (yebblies) [inline src/backend/oper.d]: Indent is wrong.

**Comment by yebblies** [comment]
Auto-merge toggled on

**Pull request merged**